### ISSUE TITLE
copyFromLocal not implemented?

### ISSUE THREAD

**Issue by interskh** [spotify/snakebite#37]
I notice copyFromLocal exists in commandlineparser.py but not in client.py. Is it not implemented yet?

Thanks!

**Comment by DandyDev** [comment]
This no word on this?
If communicating through protobuf makes it hard to implement features that require direct access to datanodes (such as the `put` and `append` operations), it would be wise to have a look at [WebHDFS](http://hadoop.apache.org/docs/r2.2.0/hadoop-project-dist/hadoop-hdfs/WebHDFS.html). Using WebHDFS in Snakebite, instead of Protobuf would make it trivial to implement `copyFromLocal`/`put`, and other file write operations.

**Comment by wouterdebie** [comment]
@ravwojdyla and I have been discussing this and currently there doesn't seem to be much time to implement this, so it's very hard to give any ETA on this feature.
I don't think we want to add WebHDFS support, since that sort of defeats the purpose of snakebite and requires additional infrastructure.

**Comment by simonellistonball** [comment]
I agree with @wouterdebie webhdfs wouldn't have the speed of snakebite. I'm working on implementing put in RPC at the moment, if anyone has any thoughts or progress they can share to accelerate it would be great to work together.

**Comment by DandyDev** [comment]
Where can I find the RPC documentation?

**Comment by zachmullen** [comment]
Has there been progress toward implementing `put`? I was going to take a crack at it for a project I'm working on, and was considering contributing it upstream, but don't want to duplicate effort if someone already has a handle on this.

**Comment by Tarrasch** [comment]
I'm pretty sure it has not, maybe @ravwojdyla can confirm.

**Comment by ravwojdyla** [comment]
I have started working on this feature some time ago - can probably upload what I have right now (it's far from complete). That said if anyone feels like working on this problem please create issues you plan to work on, and if you need help - please ping me/us. Thanks!

**Comment by zachmullen** [comment]
@ravwojdyla I'd love to help, I started to do it but the problem that ended up blocking me was that I couldn't find documentation on what RPCs I should even call to do something like an append, and the ones I tried didn't return what they claimed in the auto-generated protobuf spec... I might be able to help with this effort if you could point me to good documentation about the protocol, but I was unable to find any in sufficient detail.

**Comment by wouterdebie** [comment]
The problem with Hadoop is that protocols are pretty badly documented. When I started snakebite, I spent a lot of time reading Hadoop code and tcpdumping to figure out what was going on...

**Comment by aman572** [comment]
is there any ETA on when will copyFromLocal/put support would be present?

**Comment by tothandor** [comment]
+1

**Comment by ligao101** [comment]
+1

**Comment by mbultrow** [comment]
+1 :)

**Comment by ctimmins** [comment]
in the mean time:

import subprocess

subprocess.check_call(['hdfs', 'dfs', '-put', '/path/to/src', 'path/to/dst'], shell=False]

**Comment by jtaryma** [comment]
+1

**Comment by jwszolek** [comment]
@ravwojdyla - is there a separate branch for that issue? Did you have a chance to push what you had already done? Thanks!

**Comment by aeroevan** [comment]
It looks like a go library similar to snakebite has started making progress on writing to hdfs:
https://github.com/colinmarc/hdfs/pull/12

**Comment by Condla** [comment]
+1

**Comment by sodul** [comment]
An alternative that is relatively snappy is to use httpfs, it is a service that provide an http interface to hdfs. We actually ended up writing our own REST API in groovy to access hdfs and the hbase shell (which has no API).

https://hadoop.apache.org/docs/current/hadoop-hdfs-httpfs/index.html

**Comment by tworec** [comment]
+1

**Comment by crorella** [comment]
:+1:

**Comment by wouterdebie** [comment]
Because it was never implemented.

**Comment by austintrombley** [comment]
+1

**Comment by philpot** [comment]
+1

**Comment by zachmullen** [comment]
Github recently added a handy new feature to avoid all the "+1" comment spam. You can now vote +1 on an issue by going to the top comment, clicking the little smile face in the upper right and then click the thumbs up.

**Comment by agrebin** [comment]
+1
I've given my smiley but, just in case :)

**Comment by cherrot** [comment]
I planned to implement a storage service for big files using `snakebite` because I really like its implementation. Sadly it didn't support saving file.

Maybe I would switch back to it when this feature has been implemented :)

**Comment by francisar** [comment]
wait for this to bbe implemented

**Comment by wesmadrigal** [comment]
This was opened 3 years ago and still not implemented...wtf?

**Comment by kaiwei0** [comment]
Use webhdfs instead.

**Comment by 9nix00** [comment]
unbelievable,3 years 😱
I am using snakebite from yesterday.

**Comment by zachmullen** [comment]
If this feature is truly critical to you, I'd suggest checking out [hdfs3](http://hdfs3.readthedocs.io/en/latest/), it's BSD licensed and implements this capability. Also supports python 3.

**Comment by 9nix00** [comment]
libhdfs3 is so hard to config on MacOSX 😓  I use webhdfs write data.

**Comment by arudyk** [comment]
3 years later...

**Comment by wouterdebie** [comment]
Honestly, I'm not sure how complaining and pointing out the obvious is
going to help getting this implemented.

Yes, this feature has been open for a very long time, but writing is a
complicated operation in HDFS. Snakebite was conceived to work around long
JVM startup times, which matters mostly for operations that you do often
and should be relatively short (ls, test, etc). In cases where you read or
write, the overhead of the JVM startup time has less impact. At Spotify we
haven't had the need to invest time in write functionality, but of course
if someone feels like it, please do so. That said, please refrain from
complaining when software is open source since people do this in their
spare time or companies invest in getting software out there.

**Comment by DandyDev** [comment]
To be fair, this feature was mentioned both in an ealier version of the documentation and the original blog post announcing Snakebite. You can't blame people interpreting this as some sort of promise :)

> Snakebite was conceived to work around long JVM startup times
Is that really all there is to it? I suggested WebHDFS before, which is just a REST API, so it doesn't have anything to do with JVM startup times, but it _does_ give you a much easier path to implementing features than the undocumented Protobuf interface.
I was told back then, that using WebHDFS "defeats the purpose of Snakebite", but now I don't see how, if the purpose is to circumvent the JVM startup times that using `hdfs dfs` incurs. WebHDFS eliminates the JVM overhead just as well as Protobuf does.

**Comment by tworec** [comment]
@DandyDev maybe this SO thread will (at least partially) explain why WebHDFS is not exactly what we want: 

http://stackoverflow.com/questions/31580832/hdfs-put-vs-webhdfs

It seems that webhdfs is 4x slower.
Furthermore if we were using WebHDFS, theres no need to read Hadoop code and port it to python.

**Comment by tworec** [comment]
nice reading for this thread
http://wesmckinney.com/blog/python-hdfs-interfaces/

**Comment by baiyunping333** [comment]
I will implement the feature!

**Comment by spyzzz** [comment]
Still not implemented yet ? 
I really need this feature :|

**Comment by brunocampos01** [comment]
I suggest: https://pypi.org/project/snakebite-py3/

**Comment by derekcat** [comment]
> I suggest: https://pypi.org/project/snakebite-py3/

Why do you suggest that version?  It does not appear to have put/copyFromLocal support either.

Aeroevan's response seems to be the only thing that's actually a viable alternative at the moment:  https://github.com/colinmarc/hdfs/

Though it's a bit unclear what exactly is required as opposed to the ~/.snakebiterc (straight copies of the `core-site.xml` and `hdfs-site.xml` files from a NameNode in HDFS?), and unfortunate that it's not available as a package (doubly unfortunate for me since I'm trying to add some sort of HDFS client to all our developer utility boxes).
